### DISCUSSION TITLE
Telit LN920 development using modem-manager plugin

### DISCUSSION THREAD

**Discussion by clobrano** [fwupd/fwupd#4674] [Q&A]
Hi,

I am working to let fwupd work on Telit LN920 modem, and since it supports upgrade via MBIM, I already implemented in a local branch the required firmware interface in ModemManager and tested succesfully an upgrade via fwupdtool `install-blob` as first step.

The problem is that a similar downgrade test is failing with error message "InvalidParameters" issued by `fu_mbim_qdu_updater_file_write_ready` (so by the modem) at 93% of the writing process, which I am not sure how to debug. The stream binary was tested succesfully on Windows Update, so I think I can exclude the stream.

**Comment by hughsie** [comment]
Hmm, sounds unusual. Can you share a `--verbose` log here? Is there any timing parameter being violated, i.e. is it *always* 93%? What chunk does the opertion fail for? e.g. ctx->chunk_sent out of ctx->chunks->len.

**Reply by clobrano** [reply]
Here is the log
[fwupd-downgrade-fail93p.log](https://github.com/fwupd/fwupd/files/8770727/fwupd-downgrade-fail93p.log)

yes, it is always 93% (this is why I thought the binary was bugged, but it has been tested in WU, and should work) and I modified the log to show the chunk (76 out of 77)

**Reply by hughsie** [reply]
The...

    11:37:58:0304 FuDeviceLocker       fu_device_locker_close: assertion 'error == NULL || *error == NULL' failed

is _super sus_. Can you set `export G_DEBUG=fatal-criticals` and then get a backtrace on that critical warning please.

**Reply by clobrano** [reply]
I must say that this assertion error occurs also in the successful upgrade, so maybe it's just a wrong bit from my ModemManager branch.

**Reply by clobrano** [reply]
here is the backtrace
```
#0  0x00007f37afa62295 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#1  0x00007f37afa63579 in g_logv () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007f37afa63743 in g_log () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#3  0x00007f37a981e6f0 in fu_mm_device_finalize (object=0x55c293bc0830) at ../plugins/modem-manager/fu-mm-device.c:1830
#4  0x00007f37afb4cd0e in g_object_unref () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#5  0x00007f37a9817f1a in glib_autoptr_clear_GObject (_ptr=0x55c293bc0830) at /usr/include/glib-2.0/gobject/gobject-autocleanups.h:27
#6  0x00007f37a9817fbc in glib_autoptr_clear_FwupdDevice (_ptr=0x55c293bc0830) at ../libfwupd/fwupd-device.h:17
#7  0x00007f37a981801e in glib_autoptr_clear_FuDevice (_ptr=0x55c293bc0830) at ../libfwupdplugin/fu-device.h:19
#8  0x00007f37a981815c in glib_autoptr_clear_FuMmDevice (_ptr=0x55c293bc0830) at ../plugins/modem-manager/fu-mm-device.h:16
#9  0x00007f37a981817a in glib_autoptr_cleanup_FuMmDevice (_ptr=0x7ffde3ef5d40) at ../plugins/modem-manager/fu-mm-device.h:16
#10 0x00007f37a9818ad7 in fu_plugin_mm_inhibit_device (plugin=0x55c293b2aee0, device=0x55c293bc04f0, error=0x7ffde3ef5e10) at ../plugins/modem-manager/fu-plugin-modem-manager.c:207
#11 0x00007f37a9819533 in fu_plugin_mm_detach (plugin=0x55c293b2aee0, device=0x55c293bc04f0, progress=0x55c293b73500, error=0x7ffde3ef5e10) at ../plugins/modem-manager/fu-plugin-modem-manager.c:465
#12 0x00007f37afdef9d1 in fu_plugin_runner_device_generic_progress
    (self=0x55c293b2aee0, device=0x55c293bc04f0, progress=0x55c293b73500, symbol_name=0x7f37afe1afdd "fu_plugin_detach", device_func=0x7f37a98194a4 <fu_plugin_mm_detach>, error=0x7ffde3ef60d8)
    at ../libfwupdplugin/fu-plugin.c:934
#13 0x00007f37afdf02f9 in fu_plugin_runner_detach (self=0x55c293b2aee0, device=0x55c293bc04f0, progress=0x55c293b73500, error=0x7ffde3ef60d8) at ../libfwupdplugin/fu-plugin.c:1239
#14 0x000055c292a19962 in fu_engine_detach
    (self=0x55c293aaa820, device_id=0x55c293bc8260 "4131804b3a39e8b423e4bb49d432f369f6e43dd3", progress=0x55c293b73500, feature_flags=FWUPD_FEATURE_FLAG_NONE, error=0x7ffde3ef60d8) at ../src/fu-engine.c:2710
#15 0x000055c292a1a8b5 in fu_engine_install_blob
    (self=0x55c293aaa820, device=0x55c293bc04f0, blob_fw=0x55c293a9c240, progress=0x55c293a881e0, flags=FWUPD_INSTALL_FLAG_NO_HISTORY, feature_flags=FWUPD_FEATURE_FLAG_NONE, error=0x7ffde3ef60d8)
    at ../src/fu-engine.c:3023
#16 0x000055c2929fafb9 in fu_util_install_blob (priv=0x55c293a82520, values=0x55c293ab3c90, error=0x7ffde3ef60d8) at ../src/fu-tool.c:985
#17 0x000055c292a05b85 in fu_util_cmd_array_run (array=0x55c293a82600, priv=0x55c293a82520, command=0x7ffde3ef8259 "install-blob", values=0x7ffde3ef6618, error=0x7ffde3ef60d8) at ../src/fu-util-common.c:633
#18 0x000055c292a02c23 in main (argc=3, argv=0x7ffde3ef6608) at ../src/fu-tool.c:3771
```

backtrace full attached as file
[gdb-backtrace-full.txt](https://github.com/fwupd/fwupd/files/8772351/gdb-backtrace-full.txt)

**Reply by hughsie** [reply]
Was that for the failure to downgrade? That's a different issue to the first one in the logs (`g_object_unref: assertion 'G_IS_OBJECT (object)' failed`), further down a bit. Can you try with the wip/hughsie/modem-manager branch and try to get the backtrace again please.

**Reply by clobrano** [reply]
uhm you are right, I did not notice it was different. I will retry tomorrow since I don't have the modem available right now, and I'll checkout the branch

**Reply by clobrano** [reply]
that's weird, no matter if I `git fetch --all`, `pull` whatever, but I can't see the wip/hughsie/modem-manager branch listed.
I can apply the same patch however

**Reply by hughsie** [reply]
Weird indeed. Let me know how it goes.

**Reply by clobrano** [reply]
Tested again with your commit `53145ad5b1a0baea32d72a11b771cda2ddba13c8` applied, here is the backtrace with G_DEBUG set
```
#0  0x00007f7259947295 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#1  0x00007f7259948579 in g_logv () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007f7259948743 in g_log () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#3  0x00007f7259cb8d51 in fu_device_locker_close (self=0x55d2986f1ac0, error=0x7ffe30cc2b18) at ../libfwupdplugin/fu-device-locker.c:85
#4  0x00007f724f52ef30 in fu_mm_device_write_firmware_mbim_qdu (device=0x55d2986ec350, fw=0x55d2985c9320, progress=0x55d29869dd70, error=0x7ffe30cc2b18) at ../plugins/modem-manager/fu-mm-device.c:1252
#5  0x00007f724f52fe1c in fu_mm_device_write_firmware (device=0x55d2986ec350, firmware=0x55d2986ea380, progress=0x55d29869dd70, flags=FWUPD_INSTALL_FLAG_NO_HISTORY, error=0x7ffe30cc2b18)
    at ../plugins/modem-manager/fu-mm-device.c:1636
#6  0x00007f7259cc2508 in fu_device_write_firmware (self=0x55d2986ec350, fw=0x55d2985c9320, progress=0x55d29869dd70, flags=FWUPD_INSTALL_FLAG_NO_HISTORY, error=0x7ffe30cc2b18)
    at ../libfwupdplugin/fu-device.c:3851
#7  0x00007f7259cd6961 in fu_plugin_device_write_firmware (self=0x55d298654ee0, device=0x55d2986ec350, fw=0x55d2985c9320, progress=0x55d29869dd70, flags=FWUPD_INSTALL_FLAG_NO_HISTORY, error=0x7ffe30cc2b18)
    at ../libfwupdplugin/fu-plugin.c:726
#8  0x00007f7259cd97c0 in fu_plugin_runner_write_firmware (self=0x55d298654ee0, device=0x55d2986ec350, blob_fw=0x55d2985c9320, progress=0x55d29869dd70, flags=FWUPD_INSTALL_FLAG_NO_HISTORY, error=0x7ffe30cc2b18)
    at ../libfwupdplugin/fu-plugin.c:1949
#9  0x000055d29799fd79 in fu_engine_write_firmware
    (self=0x55d2985d8810, device_id=0x55d2986f8550 "4131804b3a39e8b423e4bb49d432f369f6e43dd3", blob_fw=0x55d2985c9320, progress=0x55d29869dd70, flags=FWUPD_INSTALL_FLAG_NO_HISTORY, error=0x7ffe30cc2b18)
    at ../src/fu-engine.c:2885
#10 0x000055d2979a04fd in fu_engine_install_blob
    (self=0x55d2985d8810, device=0x55d2986ec350, blob_fw=0x55d2985c9320, progress=0x55d2985b52e0, flags=FWUPD_INSTALL_FLAG_NO_HISTORY, feature_flags=FWUPD_FEATURE_FLAG_NONE, error=0x7ffe30cc2b18)
    at ../src/fu-engine.c:3057
#11 0x000055d2979800f2 in fu_util_install_blob (priv=0x55d2985af520, values=0x55d2985e4d50, error=0x7ffe30cc2b18) at ../src/fu-tool.c:992
#12 0x000055d29798b053 in fu_util_cmd_array_run (array=0x55d2985af600, priv=0x55d2985af520, command=0x7ffe30cc460a "install-blob", values=0x7ffe30cc3058, error=0x7ffe30cc2b18) at ../src/fu-util-common.c:634
#13 0x000055d2979880af in main (argc=3, argv=0x7ffe30cc3048) at ../src/fu-tool.c:3856
```

full backtrace 
[fu-device-locker-close-bt-full.txt](https://github.com/fwupd/fwupd/files/8777584/fu-device-locker-close-bt-full.txt)

**Reply by aleksander0m** [reply]
Looks like the given `GError` is already initialized with an error, so `fu_device_locker_close()` complains.

**Reply by clobrano** [reply]
is the one below intended to avoid overriding the error passed as argument?

[g_return_val_if_fail(error == NULL || *error == NULL, FALSE);](https://github.com/fwupd/fwupd/blob/41f5b7b563058e790ed7b20463522fe4ab1d871c/libfwupdplugin/fu-device-locker.c#L85-L86)

if that so, it seems ok

**Reply by aleksander0m** [reply]
Indeed:
```
	fu_mbim_qdu_updater_write(self->mbim_qdu_updater, filename, data, device, progress, error);
	if (!fu_device_locker_close(locker, error))
		return FALSE;
```

If `fu_mbim_qdu_updater_write()` fails and sets the error, we cannot pass the same error to `fu_device_locker_close()`. that's something to fix.

**Reply by aleksander0m** [reply]
@hughsie What's the purpose of the explicit `fu_device_locker_close()` call there? I have no idea

**Reply by hughsie** [reply]
I guess it's to close the device so that fu_mm_device_get_firmware_version_mbim() succeeds; Jarvis wrote that code in 6b79e9201d9ca7ea56ee6fe899036edbbd056991 -- I'll push a patch to avoid the leak and to capture the GError.

**Reply by hughsie** [reply]
@clobrano can you pull the latest branch and test again? I don't think that will fix your failed download, but it should fix the critical warning and return the correct error message at least.

**Reply by clobrano** [reply]
yes, I can confirm that the critical warning is gone (I can comment on the PR as well)

**Comment by hughsie** [comment]
@aleksander0m any ideas?

**Reply by aleksander0m** [reply]
let me check

**Comment by aleksander0m** [comment]
@clobrano is this with a custom branch of yours or with the main upstream branch?

**Reply by clobrano** [reply]
hey @aleksander0m, this is a local branch which implements firmware interface for telit broadband modem mbim. I can push a MR to modem-manager if you like to have a look

**Reply by clobrano** [reply]
also, if you have any idea about how to debug why `fu_mbim_qdu_updater_file_write_ready` is returning error at the last chunk would be very helpful :)
The delta firmware is expected to be ok, since it's been tested in WU

**Reply by hughsie** [reply]
@aleksander0m might know what `InvalidParameters` is trying to tell us.

**Reply by aleksander0m** [reply]
looks like I missed this followup :sweat_smile:  sorry

**Reply by aleksander0m** [reply]
As @clobrano points out, this may be relevant https://gitlab.freedesktop.org/mobile-broadband/libmbim/-/merge_requests/146

**Reply by clobrano** [reply]
> this may be relevant https://gitlab.freedesktop.org/mobile-broadband/libmbim/-/merge_requests/146

:smile: Indeed!
I understood that the `InvalidParameters` error was due to a mismatch between the expected packet size and the actual write, but (before this morning) I did not know why :+1:

I tested the process with the Libmbim MR and everything works almost perfectly
